File: container-jersey2/src/main/java/com/yahoo/container/servlet/jersey/ComponentGraphProvider.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.servlet.jersey;

import com.yahoo.container.di.config.ResolveDependencyException;
import com.yahoo.container.di.config.RestApiContext;
import com.yahoo.container.jaxrs.annotation.Component;
import org.glassfish.hk2.api.Injectee;
import org.glassfish.hk2.api.InjectionResolver;
import org.glassfish.hk2.api.ServiceHandle;

import javax.inject.Singleton;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Resolves jdisc container components for jersey 2 components.
 *
 * @author Tony Vaagenes
 * @author ollivir
 */
@Singleton // jersey2 requirement: InjectionResolvers must be in the Singleton scope
public class ComponentGraphProvider implements InjectionResolver<Component> {
    private Collection<RestApiContext.Injectable> injectables;

    public ComponentGraphProvider(Collection<RestApiContext.Injectable> injectables) {
        this.injectables = injectables;
    }

    @Override
    public Object resolve(Injectee injectee, ServiceHandle<?> root) {
        Class<?> wantedClass;
        Type type = injectee.getRequiredType();
        if (type instanceof Class) {
            wantedClass = (Class<?>) type;
        } else {
            throw new UnsupportedOperationException("Only classes are supported, got " + type);
        }

        List<RestApiContext.Injectable> componentsWithMatchingType = new ArrayList<>();
        for (RestApiContext.Injectable injectable : injectables) {
            if (wantedClass.isInstance(injectable.instance)) {
                componentsWithMatchingType.add(injectable);
            }
        }

        if (componentsWithMatchingType.size() == 1) {
            return componentsWithMatchingType.get(0).instance;
        } else {
            String injectionDescription = "class '" + wantedClass + "' to inject into Jersey resource/provider '"
                    + injectee.getInjecteeClass() + "')";
            if (componentsWithMatchingType.size() > 1) {
                String ids = componentsWithMatchingType.stream().map(c -> c.id.toString()).collect(Collectors.joining(","));
                throw new ResolveDependencyException("Multiple components found of " + injectionDescription + ": " + ids);
            } else {
                throw new ResolveDependencyException("Could not find a component of " + injectionDescription + ".");
            }
        }
    }

    @Override
    public boolean isMethodParameterIndicator() {
        return true;
    }

    @Override
    public boolean isConstructorParameterIndicator() {
        return true;
    }
}

File: container-jersey2/src/main/scala/com/yahoo/container/servlet/jersey/ComponentGraphProvider.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.servlet.jersey

import javax.inject.Singleton

import com.yahoo.container.di.config.{ResolveDependencyException, RestApiContext}
import com.yahoo.container.jaxrs.annotation.Component
import org.glassfish.hk2.api.{ServiceHandle, Injectee, InjectionResolver}

/**
 * Resolves jdisc container components for jersey 2 components.
 * Similar to Gjoran's ComponentGraphProvider for jersey 1.
 * @author tonytv
 */
@Singleton //jersey2 requirement: InjectionResolvers must be in the Singleton scope
class ComponentGraphProvider(injectables: Traversable[RestApiContext.Injectable]) extends InjectionResolver[Component] {
  override def resolve(injectee: Injectee, root: ServiceHandle[_]): AnyRef = {
    val wantedClass = injectee.getRequiredType match {
      case c: Class[_] => c
      case unsupported => throw new UnsupportedOperationException("Only classes are supported, got " + unsupported)
    }

    val componentsWithMatchingType = injectables.filter{ injectable =>
        wantedClass.isInstance(injectable.instance) }

    val injectionDescription =
      s"class '$wantedClass' to inject into Jersey resource/provider '${injectee.getInjecteeClass}')"

    if (componentsWithMatchingType.size > 1)
      throw new ResolveDependencyException(s"Multiple components found of $injectionDescription: " +
        componentsWithMatchingType.map(_.id).mkString(","))

    componentsWithMatchingType.headOption.map(_.instance).getOrElse {
      throw new ResolveDependencyException(s"Could not find a component of $injectionDescription.")
    }
  }

  override def isMethodParameterIndicator: Boolean = true
  override def isConstructorParameterIndicator: Boolean = true
}
